### PR TITLE
fix get partition metadata problem for a non-existed topic

### LINKED ISSUE #8813: Get partition metadata for a non-existed topic returns 0
**Describe the bug**
When getting the partition metadata for a non-existed topic, it returns 0 instead of throwing an exception.

**To Reproduce**
1. Run a pulsar standalone
2. Get a topic's partition number

```java
        try (PulsarAdmin admin = PulsarAdmin.builder().serviceHttpUrl("http://localhost:8080").build()) {
            System.out.println(admin.topics().getPartitionedTopicMetadata("xxxxx").partitions);
        } catch (/* ... */) {
```

3. The output is 0 and no exception was thrown.

**Expected behavior**
It should throw an exception because for a non-partitioned topic, its partitions should be 0, so we cannot distinguish a non-existed topic from a non-partitioned topic.

### PR DESCRIPTION
**Pull request by aloyszhang**
Fixes #8813

### Motivation

Currently, getting the partition metadata for a non-existed topic, it returns 0 instead of throwing an exception.
This pr fix this by  throwing an exception.

### Modifications

If no metadata found in global zk, then will check whether the topic is exist, if yes, will return 0, otherwise will  throw an exception.

### CODE REVIEW

**Review by BewareMyPower — CHANGES_REQUESTED** [review]
A unit test should be added to cover the case when a PulsarAdmin tries to get the partition metadata of a non-existed topic.

**Comment by BewareMyPower** [comment]
Please rebase to the latest master since the master was broken before,

**Comment by aloyszhang** [comment]
@BewareMyPower I'll do this later.

**Comment by wolfstudy** [comment]
@BewareMyPower Please take a look this change again, thanks.

**Comment by BewareMyPower** [comment]
@aloyszhang could you rebase to latest master?

**Review by BewareMyPower — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 1]
```diff
@@ -702,6 +702,11 @@ protected static PartitionedTopicMetadata fetchPartitionedTopicMetadata(PulsarSe
             return pulsar.getBrokerService().fetchPartitionedTopicMetadataAsync(topicName).get();
         } catch (Exception e) {
             if (e.getCause() instanceof RestException) {
+                // Since CompletableFuture wrappers exception, so we will lost the Status
+                // rebuild the Status if exception message contains "Topic not exist."
+                if (e.getCause().getMessage().contains("Topic not exist.")) {
+                    throw new org.apache.pulsar.broker.web.RestException(Response.Status.NOT_FOUND, "Topic not exist.");
+                }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:709]: Could you please clarify why `throw (RestException) e.getCause()` does not work here? If e.getCause is instance of RestException, I think we can throw it directly. I might be missing something here.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 2]
```diff
@@ -715,6 +720,11 @@ protected static PartitionedTopicMetadata fetchPartitionedTopicMetadataCheckAllo
                     .get();
         } catch (Exception e) {
             if (e.getCause() instanceof RestException) {
+                // Since CompletableFuture wrappers exception, so we will lost the Status
+                // rebuild the Status if exception message contains "Topic not exist."
+                if (e.getCause().getMessage().contains("Topic not exist.")) {
+                    throw new org.apache.pulsar.broker.web.RestException(Response.Status.NOT_FOUND, "Topic not exist.");
+                }
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:727]: Same as the above comment.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -456,10 +456,17 @@ protected void internalCreateNonPartitionedTopic(boolean authoritative) {
...
+        PartitionedTopicMetadata partitionMetadata = null;
+        try {
+            partitionMetadata = getPartitionedTopicMetadata(topicName, authoritative, false);
+            if (partitionMetadata.partitions > 0) {
+                log.warn("[{}] Partitioned topic with the same name already exists {}", clientAppId(), topicName);
+                throw new RestException(Status.CONFLICT, "This topic already exists");
+            }
+        } catch (RestException restException) {
+            if (Status.NOT_FOUND.getStatusCode() != restException.getResponse().getStatus()) {
+                throw restException;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:468]: I think it's not correct handling here, If got 404 here, it means the partitioned metadata does not exist right? If the partitioned metadata does exist, why should we prevent the non-partitioned topic creation?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 4]
```diff
@@ -1174,13 +1174,19 @@ public ServiceUnitId getServiceUnitId(TopicName topicName) throws Exception {
...
+                return;
+            }
             List<String> result = Lists.newArrayList();
             for (int i = 0; i < meta.partitions; i++) {
                 result.add(topicName.getPartition(i).toString());
             }
-            return CompletableFuture.completedFuture(result);
+            future.complete(result);
         });
+        return future;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:1189]: What is the difference between these changes? Why we change the `thenCompose` to `whenComplete`?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java (hunk 1 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:709]: I'll check this PR and push later.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 5]
```diff
@@ -702,6 +702,11 @@ protected static PartitionedTopicMetadata fetchPartitionedTopicMetadata(PulsarSe
             return pulsar.getBrokerService().fetchPartitionedTopicMetadataAsync(topicName).get();
         } catch (Exception e) {
             if (e.getCause() instanceof RestException) {
+                // Since CompletableFuture wrappers exception, so we will lost the Status
+                // rebuild the Status if exception message contains "Topic not exist."
+                if (e.getCause().getMessage().contains("Topic not exist.")) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:707]: Please do not rely on strings, can we test the actual class?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java (hunk 5 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:707]: Yes, I'll change this PR and push again.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:709]: Thanks @aloyszhang

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 6]
```diff
@@ -376,6 +377,25 @@ protected void validatePartitionedTopicMetadata(String tenant, String namespace,
...
     }
 
+    protected void validateTopicExistedAndCheckAllowAutoCreation(String tenant, String namespace,
+                                                                 String encodedTopic, boolean checkAllowAutoCreation) {
+        try {
+            PartitionedTopicMetadata partitionedTopicMetadata =
+                    pulsar().getBrokerService().fetchPartitionedTopicMetadataAsync(topicName).get();
+            if (partitionedTopicMetadata.partitions < 1) {
+                if (!pulsar().getNamespaceService().checkTopicExists(topicName).get()
+                        && checkAllowAutoCreation
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:387]: Here is not covered by tests? I noticed the tests only cover the `checkAllowAutoCreation=false`, could you please help add a test for cover `checkAllowAutoCreation=true`?

Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:387]: I'll add test case to protect this part of code.

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:387]: Thanks

**Review by codelipenghui — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by aloyszhang** [comment]
@codelipenghui  It seems merge `apache/master` branch into  this pr branch makes more checks failed.

**Pull request merged**